compiler/semantic/op.go

```diff
@@ -129,6 +125,21 @@ func (a *analyzer) semFromEntity(entity ast.FromEntity, alias *ast.TableAlias, a
 	}
 }
 
+func (a *analyzer) fromCTE(entity ast.FromEntity, c *cte, alias *ast.TableAlias) (dag.Seq, schema) {
+	if slices.Contains(a.cteStack, c.ast) {
+		a.error(entity, errors.New("recursive WITH relations not currently supported"))
+		return dag.Seq{badOp()}, badSchema()
+	}
+	a.cteStack = append(a.cteStack, c.ast)
+	body, inSchema := a.semSQLPipe(c.ast.Body, nil, &ast.TableAlias{Name: c.ast.Name.Name})
+	a.cteStack = a.cteStack[:len(a.cteStack)-1]
+	seq, outSchema, err := derefSchemaWithAlias(inSchema, alias, body)
+	if err != nil {
+		a.error(alias, err)
+	}
+	return seq, outSchema
```

Comment:
Nit: inSchema and outSchema aren't very descriptive names so just say schema.

```suggestion
	body, schema := a.semSQLPipe(c.ast.Body, nil, &ast.TableAlias{Name: c.ast.Name.Name})
	a.cteStack = a.cteStack[:len(a.cteStack)-1]
	seq, schema, err := derefSchemaWithAlias(schema, alias, body)
	if err != nil {
		a.error(alias, err)
	}
	return seq, schema
```